worlds/witness/hints.py

```diff
@@ -338,105 +394,316 @@ def make_hints(world: "WitnessWorld", hint_amount: int, own_itempool: List[Item]
 
-        if not hint_pair or hint_pair[2] == 158007:  # Tutorial Gate Open
-            continue
+    # Make sure priority hints doesn't contain any hints that are already always hints.
+    priority_hints_set -= always_hints_set
 
-        always_hint_pairs[hint_pair[0]] = (hint_pair[1], True, hint_pair[2])
+    always_generator = [hint for hint in always_hints_set if hint and hint.location not in already_hinted_locations]
+    priority_generator = [hint for hint in priority_hints_set if hint and hint.location not in already_hinted_locations]
 
-    for location in always_locations:
-        hint_pair = make_hint_from_location(world, location)
-        always_hint_pairs[hint_pair[0]] = (hint_pair[1], False, hint_pair[2])
+    # Convert both hint types to list and then shuffle. Also, get rid of None and Tutorial Gate Open.
+    always_hints = sorted(always_generator, key=lambda h: h.location)
+    priority_hints = sorted(priority_generator, key=lambda h: h.location)
+    world.random.shuffle(always_hints)
+    world.random.shuffle(priority_hints)
 
-    priority_hint_pairs = dict()
+    return always_hints, priority_hints
 
-    for item in priority_items:
-        hint_pair = make_hint_from_item(world, item, own_itempool)
 
-        if not hint_pair or hint_pair[2] == 158007:  # Tutorial Gate Open
-            continue
+def make_extra_location_hints(world: "WitnessWorld", hint_amount: int, own_itempool: List[Item],
+                              already_hinted_locations: Set[Location], hints_to_use_first: List[WitnessLocationHint],
+                              unhinted_locations_for_hinted_areas: Dict[str, Set[Location]]) -> List[WitnessWordedHint]:
+    prog_items_in_this_world, locations_in_this_world = get_items_and_locations_in_random_order(world, own_itempool)
 
-        priority_hint_pairs[hint_pair[0]] = (hint_pair[1], True, hint_pair[2])
+    next_random_hint_is_location = world.random.randrange(0, 2)
 
-    for location in priority_locations:
-        hint_pair = make_hint_from_location(world, location)
-        priority_hint_pairs[hint_pair[0]] = (hint_pair[1], False, hint_pair[2])
+    hints = []
 
-    already_hinted_locations = set()
+    # This is a way to reverse a Dict[a,List[b]] to a Dict[b,a]
+    area_reverse_lookup = {v: k for k, l in unhinted_locations_for_hinted_areas.items() for v in l}
 
-    for loc, item in always_hint_pairs.items():
-        if loc in already_hinted_locations:
+    while len(hints) < hint_amount:
+        if not prog_items_in_this_world and not locations_in_this_world and not hints_to_use_first:
+            player_name = world.multiworld.get_player_name(world.player)
+            logging.warning(f"Ran out of items/locations to hint for player {player_name}.")
+            break
+        if hints_to_use_first:
+            location_hint = hints_to_use_first.pop()
+        elif next_random_hint_is_location or not prog_items_in_this_world:
```

Comment:
This can technically crash (IndexError: pop index out of range) if `prog_items_in_this_world` is true, `next_random_hint_is_location` is true, and `locations_in_this_world` is empty.